Split out the steps (#22)
nearform/the-fastify-workshop: deck.md

# Step 2: Solution /3

```js
// routes/users.js
export default async function users(fastify) {
  fastify.get('/users', {}, async () => [
    { username: 'alice' },
    { username: 'bob' },
  ])
}
```

---

# Step 2: Trying it out

#### Note that the / route is now not found

```a
curl http://localhost:3000/

> {"message":"Route GET:/ not found","error":"Not Found","statusCode":404}

Reviewer: format it so it's readable without the scroll

![Screenshot 2021-02-02 at 16 16 45](https://user-images.githubusercontent.com/272483/106620663-1355cf80-6572-11eb-9ce4-90979e9a708c.png)

```

#### We'll find our response at /users

```a
curl http://localhost:3000/users
[{"username":"alice"},{"username":"bob"}]
```

---

class: branded

# Step 4: Serialization

- Fastify uses a schema-based approach, and even if it is not mandatory we recommend using JSON Schema to validate your routes and serialize your outputs. Internally, Fastify compiles the schema into a highly performant function

- We encourage you to use an output schema, as it can drastically increase throughput and help prevent accidental disclosure of sensitive information

https://www.fastify.io/docs/latest/Validation-and-Serialization/#serialization

---

class: branded

# Step 4: Exercise 💻

- respond to requests with an array of two users:

```json
[{ "username": "alice" }, { "username": "bob" }]

> ```

Reviewer: I think we are already doing this in step 2, we should remove it from here and place it in step 2

---

> # Step 4: Trying it out

Reviewer: was this somewhere before? Or you added it?

Author: I added this

I think each step should have a 'try it out' step so the class can see what should be happening after each change. I think this also works nicely for steps which should still work the same (step 8)

#### Make the response invalid

* In routes/users.js change the hardcoded response so it doesn't match the schema:

```javascript
async () => [{ name: 'alice' }, { name: 'bob' }]
```

> * You will need to restart in step-4 for these changes to take effect

Reviewer: `You will need to restart the server [...]` ?

---

# Step 5: Trying it out

#### Run the tests

```a
yarn test

$ tap
 PASS  test/index.test.js 2 OK 88.011ms
 PASS  test/login.test.js 5 OK 121.376ms

🌈 SUMMARY RESULTS 🌈

Suites:   2 passed, 2 of 2 completed
Asserts:  7 passed, of 7
Time:     721.846ms
----------|----------|----------|----------|----------|-------------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------|----------|----------|----------|----------|-------------------|
All files |        0 |        0 |        0 |        0 |                   |
----------|----------|----------|----------|----------|-------------------|
✨  Done in 2.85s.
```

---

class: branded

# Step 6: Validation

> - Route validation internally relies upon [Ajv](https://www.npmjs.com/package/ajv), which is a high-performance JSON Schema validator

Reviewer: not sure we should add `JWT_SECRET` in the `Validation` step 🤔

Reviewer: ![Screenshot 2021-02-02 at 16 27 31](https://user-images.githubusercontent.com/272483/106622129-91ff3c80-6573-11eb-8ba4-6bf20801379c.png)

Reviewer: This step seems to be more "configuration" then "validation" or routes 🤔

Author: Latest updates have resolved this

  response: {
    200: S.object()
      .prop('username', S.string().required())
      .prop('password', S.string().required())
  },
}

export default async function login(fastify) {
  fastify.post('/login', { schema }, async req => {
    const { username, password } = req.body
    return { username, password }
  })
}
```

---

# Step 6: Trying it out

#### With right credentials

```a
curl -X POST -H "Content-Type: application/json" \
-d '{ "username": "alice", "password": "alice" }'
http://localhost:3000/login
```

```json
{
  "username": "alice",
  "password": "alice"
}
```

---

# Step 6: Trying it out /2

#### With wrong credentials

```a
curl -X POST -H "Content-Type: application/json" \
-d '{ "name": "alice", "passcode": "alice" }'
http://localhost:3000/login
```

```json
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "body should have required property 'username'"
}
```

---

class: branded

# Step 7: Authentication

- [`fastify-jwt`](https://github.com/fastify/fastify-jwt) contains JWT utils for Fastify, internally uses [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken).

---

class: branded

# Step 7: Exercise 💻

- Change `index.js` so that it:

>   - Registers the `fastify-jwt` plugin using the configuration option `JWT_SECRET` as the `secret` property of the plugin's configuration options.

Reviewer: in this step we do not have yet the configuration

class: branded

# Step 8: Config

> - It is preferable to use a config file (.env) to keep track of environment variables used in the system, for example, the JWT secret we put directly into the code in the previous step.

Reviewer: ```suggestion
- It is preferable to use environment variables to configure your app, for example, the JWT secret we put directly into the code in the previous step.
```

> - This makes it easier to deploy the same code into different environments as we only need to change the config file rather than having to deploy different copies of the code.

Reviewer: ```suggestion
- This makes it easier to deploy the same code into different environments.
```

> - Typically config would not be committed to a repository and would be managed in different environments accordingly. For example, locally, developers might want to see everything from the log, whereas in Production we wouldn't want to generate huge log files and might restrict the log, outputting just errors.

Reviewer: ```suggestion
- Typically config values are not committed to a repository and they are managed with environment variables. An example would be the logging level: in production it's usually better to have only errors, while in a dev environment it may be useful to show more.
```
